internal/yunikorn/service.go

```diff
@@ -22,22 +23,28 @@ type Service struct {
 	syncInterval time.Duration
 	// eventCollectorRunning is a flag that indicates whether the event collector is running.
 	eventCollectorRunning bool
+	// workqueue processes jobs which store data in database during data sync and retries them with exponential backoff.
+	workqueue *workqueue.WorkQueue
 }
 
-func NewService(repository repository.Repository, eventRepository repository.EventRepository, client Client) *Service {
+func NewService(ctx context.Context, repository repository.Repository, eventRepository repository.EventRepository, client Client) *Service {
 	s := &Service{
 		repo:            repository,
 		eventRepository: eventRepository,
 		client:          client,
 		appMap:          make(map[string]*dao.ApplicationDAOInfo),
 		syncInterval:    5 * time.Minute,
+		workqueue:       workqueue.NewWorkQueue(workqueue.WithName("yunikorn_data_sync")),
 	}
 	s.eventHandler = s.handleEvent
+	go s.workqueue.Run(ctx)
+
 	return s
 }
 
 func (s *Service) Shutdown() {
 	s.eventCollectorRunning = false
+	s.workqueue.Shutdown()
```

Comment:
Shouldn't this be before the previous line (`s.eventCollectorRunning = false`)?  Are we checking `evenCollectorRunning` somewhere far away?

Reply:
refactored a bit